### ISSUE TITLE
Refresh token stopped working and trying to re-generate it fails.

### ISSUE THREAD

**Issue by AdrianGarside** [tsightler/ring-mqtt#243]
![image](https://user-images.githubusercontent.com/26147576/131232781-168ad2ed-a89f-4fa7-ac77-b958411fd584.png)

It looks like the instructions have changed some since the last time the token stopped working across a pi reboot. Since the other options require installing stuff I don't currently have I haven't tried them yet. Just the ring-auth-cli.js one which is erroring out:

pi@pi:/etc/ring-mqtt $ docker run -it --rm --entrypoint /app/ring-mqtt/node_modules/ring-client-api/ring-auth-cli.js tsightler/ring-mqtt

#
# Fatal error in , line 0
# unreachable code
#
#
#
#FailureMessage Object: 0xbec1a22c

**Comment by tsightler** [comment]
This is most likely a duplicate of #237 and caused by the update of the container image to Alpine 3.14.  Please follow the instructions there to update the outdated Docker components on your distro.

**Comment by AdrianGarside** [comment]
I'm hitting an error at the first step:
pi@pi:~ $ sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 04EE7237B7D453EC 648ACFD622F3D138
Executing: /tmp/apt-key-gpghome.L1gE72zmDw/gpg.1.sh --keyserver keyserver.ubuntu.com --recv-keys 04EE7237B7D453EC 648ACFD622F3D138
gpg: keyserver receive failed: No such file or directory

I've been googling but not found a solution for this error signature yet.

**Comment by tsightler** [comment]
It's probably just that the key has changed.  I'd try just skipping that step, I think it will just prompt you to accept the key the first time you run apt update.

**Comment by tsightler** [comment]
Also, absolute worst case you can just manually download and force install (it's possible the repos are in transition since Debian recently release bullseye so buster is no longer current).

```
wget http://ftp.debian.org/debian/pool/main/libs/libseccomp/libseccomp-dev_2.4.4-1~bpo10+1_armhf.deb
sudo dpkg libseccomp-dev_2.4.4-1~bpo10+1_armhf.deb
```

I actually just installed the keys on one of my debian installs and they worked fine, so I'm not really sure what the problem is, but I don't have anything left running Raspbian/Raspberry PI OS as I've given up running perpetually outdated 32-bit distros, so I'm unlikely to be much help.  The other option is to install Docker from the official Docker repo instead of the super out-of-date Docker in the Raspbian repos.

**Comment by AdrianGarside** [comment]
I couldn't work out how to get apt to allow an insecure repository (the apparent option to do so didn't work in this case) but that would be bad anyway. But I finally found someone else that hit this and found a/the solution and it worked for me:
https://bbs.archlinux.org/viewtopic.php?id=252494

That got me unblocked and I was then successfully able to follow the instructions to generate a new ring token but it gave me the same value that was already in etc/ring-mqtt/ring-state.json (unless running the script already updated that but it seems not from the timestamp which is July 30th).

But I'm still hitting the same failure at container startup:

pi@pi:~ $ sudo docker attach ringmqtt
2021-08-29T17:54:08.936Z ring-mqtt Using configuration file: /data/config.json
2021-08-29T17:54:08.941Z ring-mqtt Configuration file not found, attempting to use environment variables for configuration.
2021-08-29T17:54:08.944Z ring-mqtt Reading latest data from state file: /data/ring-state.json
2021-08-29T17:54:09.154Z ring-mqtt Attempting connection to Ring API using saved refresh token.
2021-08-29T17:54:09.573Z ring-mqtt Refresh token is not valid.  Unable to authenticate with Ring servers.  See https://github.com/dgreif/ring/wiki/Refresh-Tokens
2021-08-29T17:54:09.573Z ring-mqtt Unable to connect to Ring API using saved refresh token.
2021-08-29T17:54:09.574Z ring-mqtt Attempting connection to Ring API using RINGTOKEN environment variable.
2021-08-29T17:54:09.710Z ring-mqtt Refresh token is not valid.  Unable to authenticate with Ring servers.  See https://github.com/dgreif/ring/wiki/Refresh-Tokens
2021-08-29T17:54:09.710Z ring-mqtt Could not create the API instance. This could be because the Ring servers are down/unreachable
2021-08-29T17:54:09.711Z ring-mqtt or maybe all available refresh tokens are invalid.
2021-08-29T17:54:09.711Z ring-mqtt Please check the configuration and network settings, or generate a new refresh token, and try again.
2021-08-29T17:54:09.712Z ring-mqtt Exit code: 2

**Comment by AdrianGarside** [comment]
I was wrong above. The new token was _almost_ identical but there were a few characters different in the middle. It's still failing but I'm getting a 404 error now:
2021-08-29T18:04:38.349Z ring-mqtt Using configuration file: /data/config.json
2021-08-29T18:04:38.354Z ring-mqtt Configuration file not found, attempting to use environment variables for configuration.
2021-08-29T18:04:38.357Z ring-mqtt Reading latest data from state file: /data/ring-state.json
2021-08-29T18:04:38.580Z ring-mqtt Attempting connection to Ring API using saved refresh token.
**2021-08-29T18:04:39.744Z ring-mqtt Response code 404 (Not Found)**
2021-08-29T18:04:39.745Z ring-mqtt Unable to connect to Ring API using saved refresh token.
2021-08-29T18:04:39.746Z ring-mqtt Attempting connection to Ring API using RINGTOKEN environment variable.
2021-08-29T18:04:39.884Z ring-mqtt Refresh token is not valid.  Unable to authenticate with Ring servers.  See https://github.com/dgreif/ring/wiki/Refresh-Tokens
2021-08-29T18:04:39.885Z ring-mqtt Could not create the API instance. This could be because the Ring servers are down/unreachable
2021-08-29T18:04:39.885Z ring-mqtt or maybe all available refresh tokens are invalid.
2021-08-29T18:04:39.886Z ring-mqtt Please check the configuration and network settings, or generate a new refresh token, and try again.
2021-08-29T18:04:39.887Z ring-mqtt Exit code: 2

**Comment by AdrianGarside** [comment]
It wasn't, I still had the earlier output higher in the bash window:

pi@pi:~ $ sudo docker attach ringmqtt
2021-08-29T17:54:08.936Z ring-mqtt Using configuration file: /data/config.json
2021-08-29T17:54:08.941Z ring-mqtt Configuration file not found, attempting to use environment variables for configuration.
2021-08-29T17:54:08.944Z ring-mqtt Reading latest data from state file: /data/ring-state.json
2021-08-29T17:54:09.154Z ring-mqtt Attempting connection to Ring API using saved refresh token.
2021-08-29T17:54:09.573Z ring-mqtt Refresh token is not valid.  Unable to authenticate with Ring servers.  See https://github.com/dgreif/ring/wiki/Refresh-Tokens
2021-08-29T17:54:09.573Z ring-mqtt Unable to connect to Ring API using saved refresh token.
2021-08-29T17:54:09.574Z ring-mqtt Attempting connection to Ring API using RINGTOKEN environment variable.
2021-08-29T17:54:09.710Z ring-mqtt Refresh token is not valid.  Unable to authenticate with Ring servers.  See https://github.com/dgreif/ring/wiki/Refresh-Tokens
2021-08-29T17:54:09.710Z ring-mqtt Could not create the API instance. This could be because the Ring servers are down/unreachable
2021-08-29T17:54:09.711Z ring-mqtt or maybe all available refresh tokens are invalid.
2021-08-29T17:54:09.711Z ring-mqtt Please check the configuration and network settings, or generate a new refresh token, and try again.
2021-08-29T17:54:09.712Z ring-mqtt Exit code: 2

**Comment by tsightler** [comment]
I have no choice but to believe the error message that the token isn't valid, as ring-client-api is used by 1000's of users at this point between ring-mqtt and the ring-homebridge addons, and no one else has reported any issues generating tokens.  Are you sure you copied the token correctly when you updated the RINGTOKEN environment variable because that's the only thing I could think of, or did you try to update the ring-state.json directly?

Also, just to be sure, you are definitely on the latest 4.7.3 version, correct?  The 404 made me suspicious because that is a known issue with any version older the 4.6.2 since Ring made changes to the API those version will always fail with 404.  I assume it must be 4.7.3 because that's the only version on Alpine 3.14, but I wasn't sure since I don't see anything in the log.

**Comment by AdrianGarside** [comment]
I was updating ring-state.json directly. Is that not supported? I can run the the container manually with the new token on the command line and it successfully got past the ring login and updated the ring-state.json file (adding a new systemId line) with a modified token. It then got blocked because apparently I changed my mqttbroker password since I stored the container run command so I have to work out what I changed it to. Restarting the stored docker container still gives me the same 404 error though so the updated ring-state.json content wasn't sufficient.

How do I tell what version I'm running in my container? I think I updated the container yesterday but it's likely I didn't do that/do that properly given what you just said. 

Edit: Yes, looks like that's the case I pulled a new image but the container is still running the 10 month old one. Running it manually created a fresh container so I just need to fix the incorrect mqtt broker password and I think I'll be back up and running.

**Comment by AdrianGarside** [comment]
--net="host" was the missing command line setting to fix the broker access
Everything is up and running again properly now.

**Comment by tsightler** [comment]
Glad you got it working and thanks for sharing what you found.  Hopefully over the coming weeks Raspberry Pi OS will finally update their docker packages, or at least release a bullseye based version of the OS, and people will start upgrading, and this will be behind us.  The first backport to address this problem was released on 8-31-2019, so two years ago!  Really no excuse that this hasn't made it's way to the base release, except, well, it's Debian so it's "stable", except in this case stable means broken packages with outdated and broken system calls which won't run modern Docker images.

It's why I finally switched my RPis to 64-bit Ubuntu as I was tired of dealing with it.  Most of the addons in HA moved to Alpine 3.13 ~6 months ago, and, at the time, I had to switch my Docker image from the base hassio-addon image just to keep compatibility with these broken distros, because I build the same image for both the addon and standard Docker users, but I kept having to pull in more and more manual stuff because Alpine 3.12 was getting more and more out of date.  With the upcoming addition of video streaming, it just became too much and I decided to bite the bullet and move to Apline 3.14 from the addon base image, and users with broken distros would just have to upgrade their package.  So far it hasn't been too bad but sorry it was so much trouble for you.
